src/fields/multi_field.impl.hh

```diff
@@ -150,38 +142,72 @@ void MultiField<spacedim, Value>::set_up_components() {
 }
 
 
 
 template<int spacedim, class Value>
-typename Field<spacedim,Value>::FieldBasePtr MultiField<spacedim, Value>::MultiFieldFactory::create_field(Input::Record descriptor_rec, const FieldCommon &field) {
-	Input::Record multifield_rec;
-	if (descriptor_rec.opt_val(field.input_name(), multifield_rec));
-	Input::Iterator<Input::AbstractRecord> it_common = multifield_rec.find<Input::AbstractRecord>("common");
-	Input::Iterator<Input::Array> it_components = multifield_rec.find<Input::Array>("components");
-	if (it_common && !it_components) {
-		it_common->transpose_to( multifield_rec, "components", spacedim );
-		it_components = multifield_rec.find<Input::Array>("components");
+void MultiField<spacedim,Value>::set_input_list(const Input::Array &list) {
+    if (! flags().match(FieldFlag::declare_input)) return;
+
+    this->full_input_list_ = list;
+    shared_->list_idx_ = 0;
+}
+
+
+template<int spacedim, class Value>
+std::vector<typename Value::return_type> MultiField<spacedim, Value>::value(const Point &p, const ElementAccessor<spacedim> &elm) const {
+	std::vector<typename Value::return_type> value;
+	for (auto field : sub_fields_) {
+		value.push_back( field.value(p, elm) );
 	}
+	return value;
+}
+
 
-	ASSERT(it_components, "Failed to fill 'components' array of multifield: %s.", field.input_name().c_str());
-	ASSERT(index_ < it_components->size(), "Index of MultiField component is out of range.\n");
 
-	unsigned int position = 0;
-	for (auto it = it_components->begin<Input::AbstractRecord>(); it != it_components->end(); ++it, ++position)
+template<int spacedim, class Value>
+void MultiField<spacedim, Value>::value_list(const std::vector< Point >  &point_list, const  ElementAccessor<spacedim> &elm,
+                   std::vector< std::vector<typename Value::return_type> >  &value_list) const {
+	value_list.clear();
+	for (auto field : sub_fields_) {
+		std::vector<typename Value::return_type> value;
+		field.value_list(point_list, elm, value);
+		value_list.push_back( value );
+	}
+}
+
+
+
+template<int spacedim, class Value>
+typename Field<spacedim,Value>::FieldBasePtr MultiField<spacedim, Value>::MultiFieldFactory::create_field(Input::Record descriptor_rec, const FieldCommon &field) {
+	Input::Array multifield_arr;
+	if (descriptor_rec.opt_val(field.input_name(), multifield_arr))
 	{
-		if (index_ == position) {
-			typename Field<spacedim,Value>::FieldBasePtr field_algo_base = Field<spacedim,Value>::FieldBaseType::function_factory( (*it), field.n_comp() );
-			field_algo_base->set_component_idx(index_);
-			return field_algo_base;
+		unsigned int position = 0;
+		for (auto it = multifield_arr.begin<Input::AbstractRecord>(); it != multifield_arr.end(); ++it, ++position)
+		{
+			if (index_ == position) {
+				typename Field<spacedim,Value>::FieldBasePtr field_algo_base = Field<spacedim,Value>::FieldBaseType::function_factory( (*it), field.n_comp() );
+				field_algo_base->set_component_idx(index_);
+				return field_algo_base;
+			}
```

Comment:
Yet missing check for the case multifield_arr.size() != n_components and multifield_arr.size() == 1.
Moreover we need similar check in setup_components but for the whole full_input_list.